contracts/src/BatchManagerOperations.sol

```diff
@@ -0,0 +1,608 @@
+        );
+    }
+
+    function removeFromBatch(
+        uint256 _troveId,
+        uint256 _newAnnualInterestRate,
+        uint256 _upperHint,
+        uint256 _lowerHint,
+        uint256 _maxUpfrontFee
+    ) external {
+        _removeFromBatchInternal(
+            _troveId,
+            _newAnnualInterestRate,
+            _upperHint,
+            _lowerHint,
+            _maxUpfrontFee,
+            false
+        );
+    }
+
+    function _removeFromBatchInternal(
+        uint256 _troveId,
+        uint256 _newAnnualInterestRate,
+        uint256 _upperHint,
+        uint256 _lowerHint,
+        uint256 _maxUpfrontFee,
+        bool _kick
+    ) internal {
+        LocalVariables_removeFromBatch memory vars;
+        vars.troveManager = troveManager;
+        vars.sortedTroves = sortedTroves;
+
+        if (_kick) {
+            _requireTroveIsOpen(vars.troveManager, _troveId);
+        } else {
+            _requireTroveIsActive(vars.troveManager, _troveId);
+            _requireCallerIsBorrower(_troveId);
+            _requireValidAnnualInterestRate(_newAnnualInterestRate);
+        }
+
+        vars.batchManager = _requireIsInBatch(_troveId);
+        vars.trove = vars.troveManager.getLatestTroveData(_troveId);
+        vars.batch = vars.troveManager.getLatestBatchData(vars.batchManager);
+
+        if (_kick) {
+            if (
+                vars.batch.totalDebtShares * MAX_BATCH_SHARES_RATIO >=
+                vars.batch.entireDebtWithoutRedistribution
+            ) {
+                revert BatchSharesRatioTooLow();
+            }
+            _newAnnualInterestRate = vars.batch.annualInterestRate;
+        }
+
+        IBorrowerOperations(address(this)).removeTroveFromBatch(_troveId);
+
+        if (!_checkTroveIsZombie(vars.troveManager, _troveId)) {
+            // Remove trove from Batch in SortedTroves
+            vars.sortedTroves.removeFromBatch(_troveId);
+            // Reinsert as single trove
+            vars.sortedTroves.insert(
+                _troveId,
+                _newAnnualInterestRate,
+                _upperHint,
+                _lowerHint
+            );
+        }
+
+        vars.batchFutureDebt =
+            vars.batch.entireDebtWithoutRedistribution -
+            (vars.trove.entireDebt - vars.trove.redistBoldDebtGain);
+
+        vars.batchChange.appliedRedistBoldDebtGain = vars
+            .trove
+            .redistBoldDebtGain;
+        vars.batchChange.appliedRedistCollGain = vars.trove.redistCollGain;
+        vars.batchChange.batchAccruedManagementFee = vars
+            .batch
+            .accruedManagementFee;
+        vars.batchChange.oldWeightedRecordedDebt = vars
+            .batch
+            .weightedRecordedDebt;
+        vars.batchChange.newWeightedRecordedDebt =
+            vars.batchFutureDebt *
+            vars.batch.annualInterestRate +
+            vars.trove.entireDebt *
+            _newAnnualInterestRate;
+
+        // Apply upfront fee on premature adjustments
```

Comment:
also here part of the original comment was removed, see https://github.com/mento-protocol/bold/pull/14/files#diff-d30cc87ed051c6909d6e9b48a0d719549c2272b1418110c6f55618d5ec6d9448L1434